File: ch3/rendezvous/java/Barrier.java
package Concurrency;

import java.util.Random;
import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Multiplex{
//Puzzle: Generalize the previous solution so that it allows multiple threads to run in the critical section at the same time,
// but it enforces an upper limit on the number of concurrent threads. In other words, no more than n threads can run in the
// critical section at the same time.

    static Semaphore barrier;
    static Lock mutex;
    static int count;
    static Random rand;

    static class Worker extends Thread{
        String name;
        public Worker(String name){
            this.name = name;
        }
        public void run() {
            System.out.println(name+"1");
            mutex.lock();
            count++;
            mutex.unlock();
            System.out.println(count);
            if(count == 3) barrier.release();
            long sleepTime = rand.nextInt(1000);
            try {
                Thread.sleep(sleepTime);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            try {
                barrier.acquire();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            barrier.release();
            System.out.println(name+"2");
        }
    }

    public static void main(String[] args) {
        mutex = new ReentrantLock(true);
        rand = new Random();
        for(int i = 0; i < 10; i ++){
            count = 0;
            barrier = new Semaphore(0);
            Thread a = new Worker("A");
            Thread b = new Worker("B");
            Thread c = new Worker("C");
            a.start();
            b.start();
            c.start();
            try {
                Thread.sleep(1000L);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println("--------------");
        }
    }
}

File: ch3/barrier/python/barrier.py
from threading import Thread, Semaphore
import time
from random import randint

class Person(Thread):
    def __init__(self,i,m,s,count,numthreads):
        self.s = s
        self.m = m
        self.count = count
        self.count.i = 0
        self.numthreads = numthreads
        super().__init__(name=i)

    def run(self):
        time.sleep(1e-3*randint(1,10))

        print(f"{self.name} rendez")

        # barrier
        self.m.acquire()
        self.count.i += 1
        if self.count.i == self.numthreads:
            self.s.release()
        self.m.release()
        self.s.acquire()
        self.s.release()

        print(f"{self.name} critical section")

class Count():
    pass

def main():
    nthreads = 10
    m = Semaphore(1)
    s = Semaphore(0)
    count = Count()
    thr = []
    for i in range(nthreads):
        thr.append(Person(i+1, m, s, count, nthreads))
    for t in thr:
        t.start()
    for t in thr:
        t.join()

main()
